### PR TITLE
Add Methods in imageMatching

### PR DESCRIPTION
**Pull request by theoleplomb**
## Description

Add new methods in ImagesMatching to select best candidates for FeatureMatching.
Expose new parameters in FeatureMatching and SfM.

## Features list

- [X] ImageMatching: Implement new `Sequential` method.
- [X] ImageMatching: Implement new `Frustum` intersection method.
- [x] A new `method` param allow to choose between:
-- Sequential: based on filenames order
-- Vocabulary Tree
-- Sequential and Vocabulary Tree
-- Exhaustive
-- Frustum: when cameras have known poses, compute frustum intersections
- [X] ImageMatching: Add `nbNeighbor` param in imageMatching for Sequential method.
- [X] FeatureMatching: Add `matchFromKnownCameraPoses` param
- [X] SfM: Add `filterTrackForks` param
- [X] SfM: Add `minNbMatches` param

### CODE REVIEW

File: src/aliceVision/matching/io.cpp [hunk 1]
```diff
@@ -99,10 +99,13 @@ void filterMatchesByViews(
...
 void filterTopMatches(
   PairwiseMatches & allMatches,
-  const int limitNum)
+  const int limitNum,
+  const int minNum)
 {
   if (limitNum <= 0)
     return;
+  if (minNum <= 0 || minNum > limitNum)
+    return;
```
Comment (theoleplomb) [inline src/aliceVision/matching/io.cpp]: TODO:
```
if((limitNum <= 0 || minNum > limitNum) && minNum <= 0)
    return;
```

File: src/aliceVision/matching/io.cpp [hunk 2]
```diff
@@ -111,6 +114,8 @@ void filterTopMatches(
       IndMatches& m = matches.second;
       if (m.size() > limitNum)
```
Comment (theoleplomb) [inline src/aliceVision/matching/io.cpp]: if (limitNum > 0 && m.size() > limitNum)

File: src/aliceVision/matching/io.cpp [hunk 3]
```diff
@@ -111,6 +114,8 @@ void filterTopMatches(
...
         m.erase(m.begin()+limitNum, m.end());
+      if (m.size() < minNum)
```
Comment (theoleplomb) [inline src/aliceVision/matching/io.cpp]: if (minNum > 0 && m.size() < minNum)

File: src/software/pipeline/main_computeStructureFromKnownPoses.cpp [hunk 4]
```diff
@@ -59,7 +59,10 @@ int main(int argc, char **argv)
     ("describerTypes,d", po::value<std::string>(&describerTypesName)->default_value(describerTypesName),
       feature::EImageDescriberType_informations().c_str())
     ("matchesFolders,m", po::value<std::vector<std::string>>(&matchesFolders)->multitoken()->required(),
-      "Path to folder(s) in which computed matches are stored.");
+      "Path to folder(s) in which computed matches are stored.")
+    ("knownPosesGeometricErrorMax", po::value<double>(&knownPosesGeometricErrorMax)->default_value(knownPosesGeometricErrorMax),
```
Comment (theoleplomb) [inline src/software/pipeline/main_computeStructureFromKnownPoses.cpp]: geometricErrorMax

File: src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.cpp [hunk 5]
```diff
@@ -58,23 +58,25 @@ void PointsToMat(
 }
 
 /// Use geometry of the views to compute a putative structure from features and descriptors.
-void StructureEstimationFromKnownPoses::run(
-  SfMData& sfmData,
+void StructureEstimationFromKnownPoses::run(SfMData& sfmData,
   const PairSet& pairs,
-  const feature::RegionsPerView& regionsPerView)
+  const feature::RegionsPerView& regionsPerView,
+  double knownPosesGeometricErrorMax)
```
Comment (theoleplomb) [inline src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.cpp]: geometricErrorMax

File: src/aliceVision/matching/io.cpp [hunk 6]
```diff
@@ -272,7 +278,9 @@ bool Load(
     filterMatchesByDesc(matches, descTypesFilter);
 
   if(maxNbMatches > 0)
-    filterTopMatches(matches, maxNbMatches);
+    filterTopMatches(matches, maxNbMatches, minNbMatches);
+  if(minNbMatches <= maxNbMatches)
+    filterTopMatches(matches, maxNbMatches, minNbMatches);
```
Comment (fabiencastan) [inline src/aliceVision/matching/io.cpp]: bug: duplicates

File: src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.cpp [hunk 7]
```diff
@@ -58,23 +58,25 @@ void PointsToMat(
 }
 
 /// Use geometry of the views to compute a putative structure from features and descriptors.
-void StructureEstimationFromKnownPoses::run(
-  SfMData& sfmData,
+void StructureEstimationFromKnownPoses::run(SfMData& sfmData,
   const PairSet& pairs,
-  const feature::RegionsPerView& regionsPerView)
+  const feature::RegionsPerView& regionsPerView,
+  double GeometricErrorMax)
```
Comment (fabiencastan) [inline src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.cpp]: lower case

File: src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.hpp [hunk 8]
```diff
@@ -19,18 +19,18 @@ class StructureEstimationFromKnownPoses
 public:
 
   /// Use geometry of the views to compute a putative structure from features and descriptors.
-  void run(
-    sfmData::SfMData& sfmData,
+  void run(sfmData::SfMData& sfmData,
     const PairSet& pairs,
-    const feature::RegionsPerView& regionsPerView);
+    const feature::RegionsPerView& regionsPerView,
+    double GeometricErrorMax);
```
Comment (fabiencastan) [inline src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.hpp]: lower case

File: src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.hpp [hunk 9]
```diff
@@ -19,18 +19,18 @@ class StructureEstimationFromKnownPoses
...
 public:
 
   /// Use guided matching to find corresponding 2-view correspondences
-  void match(
-    const sfmData::SfMData& sfmData,
+  void match(const sfmData::SfMData& sfmData,
     const PairSet& pairs,
-    const feature::RegionsPerView& regionsPerView);
+    const feature::RegionsPerView& regionsPerView,
+    double GeometricErrorMax);
```
Comment (fabiencastan) [inline src/aliceVision/sfm/pipeline/structureFromKnownPoses/StructureEstimationFromKnownPoses.hpp]: same

File: src/software/pipeline/main_computeStructureFromKnownPoses.cpp [hunk 10]
```diff
@@ -37,7 +37,7 @@ int main(int argc, char **argv)
   std::string sfmDataFilename;
   std::string outSfMDataFilename;
   std::vector<std::string> featuresFolders;
-
+  double GeometricErrorMax = 5.0;
```
Comment (fabiencastan) [inline src/software/pipeline/main_computeStructureFromKnownPoses.cpp]: same

File: src/software/pipeline/main_imageMatching.cpp [hunk 11]
```diff
@@ -377,21 +598,26 @@ int main(int argc, char** argv)
       "SfMData file.")
     ("featuresFolders,f", po::value<std::vector<std::string>>(&featuresFolders)->multitoken()->required(),
       "Path to folder(s) containing the extracted features.")
-    ("tree,t", po::value<std::string>(&treeName),
+    ("tree,t", po::value<std::string>(&treeName)->default_value(treeName),
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]: move to optional params

File: src/aliceVision/matching/io.cpp [hunk 12]
```diff
@@ -99,18 +99,23 @@ void filterMatchesByViews(
...
 void filterTopMatches(
   PairwiseMatches & allMatches,
-  const int limitNum)
+  const int limitNum,
+  const int minNum)
 {
-  if (limitNum <= 0)
+  if (limitNum <= 0 && minNum <=0)
     return;
+  if (minNum > limitNum)
```
Comment (fabiencastan) [inline src/aliceVision/matching/io.cpp]: (limitNum > 0 && minNum > limitNum)

**Pull request merged**